pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -315,23 +324,61 @@ func GetRegistries(ctx *types.SystemContext) ([]Registry, error) {
 
 // FindUnqualifiedSearchRegistries returns all registries that are configured
 // for unqualified image search (i.e., with Registry.Search == true).
-func FindUnqualifiedSearchRegistries(registries []Registry) []Registry {
+func FindUnqualifiedSearchRegistries(ctx *types.SystemContext) ([]Registry, error) {
+	registries, err := GetRegistries(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	unqualified := []Registry{}
 	for _, reg := range registries {
 		if reg.Search {
 			unqualified = append(unqualified, reg)
 		}
 	}
-	return unqualified
+	return unqualified, nil
+}
+
+// isPrefix checks if the specified prefix is a namespace or an image prefix of
+// reference.  The basic assumption is that the remainder of the reference is
+// either an empty string (i.e., a full match) or a string starting with '/'
+// (i.e., a namespace prefix) or a string starting with a ':' (i.e., an image
+// prefix).
+//
+// "docker.io/library/alpine" is a prefix for "docker.io/library/alpine/path"
+// and for "docker.io/library/alpine:latest" but not for "docker.io/library/alpinenext".
+func isPrefix(reference, prefix string) bool {
+	if strings.HasPrefix(reference, prefix) {
+		remainder := reference[len(prefix):]
+		// full match
+		if len(remainder) == 0 {
+			return true
+		}
+		// The regex expects strings that either start with a '/' or
+		// that start with a ':'.  Strings starting with a ':' have a
+		// subsequent lookahead for ':' and '/' to avoid matching
+		// registries at port boundaries.
+		//
+		// This way, it matches "/foobar", "/foobar:latest" and ":latest"
+		// but it doesn't match ":5000/foobar" or ":latest:foobar".
+		reg := regexp.MustCompile(`(?m)^(\/.*|:.[^:\/]*)$`)
```

Comment:
> It should be possible to ultimately implement the desired semantics, with the current inputs, exactly correctly (e.g. by exporting DomainWithOptionalName in c/image/docker/registry, or something like that [...].

I was thinking about this issue for a while, and believe we can reduce complexity by just requiring the suffix/remainder to either be empty or to start either with `/` or with `:`. This would also match at port boundaries, but that might not be such a big deal. Some users might even find a usecase for it when wanting to mirror multiple registries running on one host.

We can still introduce a `DomainWithOptionalName` regex (or something similar) for input validation, but I don't think that's necessary now can be added at a later point.

Reply:
> > Ultimately, matching on the suffix only seems insufficient, because `:5000` can either be a valid tag or a valid port number. So, `example.com:5000` should not match `example.com`, but `example.com/image:5000` should match `example.com/image`.
>
> This is already working and the regex detects that (minus the digest case).

How? AFAICS it can’t work because the suffix is the same, and testing confirms it: https://play.golang.org/p/0QHfsmaoApO .

Am I overlooking something that can distinguish the two cases?

Reply:
> I believe we can reduce complexity by just requiring the suffix/remainder to either be empty or to start either with `/` or with `:`. This would also match at port boundaries, but that might not be such a big deal. Some users might even find a usecase for it when wanting to mirror multiple registries running on one host.

I’d rather do this precisely correctly; it seems to me it should not be that much code, maybe I could try to do that next week.

(Either way, the problematic matching code already exists, so it is not strictly blocking this PR.  The skopeo test failures are — and I’m getting close to finishing the CLI rework in there, projectatomic/skopeo#523 — and that rework seems mostly unnecessary for the purpose of this PR, after all; I will know for sure after I finish the tests in there.)